### ISSUE TITLE
Debug protocol extension: modules window

### ISSUE THREAD

**Issue by gregg-miskelly** [Microsoft/vscode-debugadapter-node#29]
To be able to use the debug protocol in other IDEs (Visual Studio, Xamarin Studio) or VS Code in the future we should think about how the debug protocol should support a modules window.

VS requires module load/unload events to keep its modules window updated, so I would likely suggest we do the same.

VS bakes in knowledge of all the module properties, so while I might suggest having some sort of 'additional properties' that windows displays, we probably want to have defined names for allow of the following properties so that we can cleanly map them onto VS interfaces. Here are the properties:

    id: string
    name : string
    path: string
    isOptimized?: bool
    isUseCode?: bool
    version? : string
    symbol status?: string
    symbolFilePath?: string
    dateTimeStamp?: string (or maybe this should be a date time?)
    addressRange?: not sure what type this should be. In VS this needs to be a UINT64 for the start, and a UINT32 for the size.

In addition, we also provide:

1. Symbol search details (this feels like it shouldn't be part of the event, but should rather be a request)
2. A command to load symbols for a specific module

**Comment by weinand** [comment]
@gregg-miskelly I think that we should not bake the concept of 'modules' and their property names into the protocol because that would make it difficult to reuse the module window for debuggers that do not have modules. E.g. chrome or node call these things 'scripts' and they would have very different properties. 

I suggest that this module window is basically a configurable table (grid) that shows objects with properties. The debug adapter returns the labels (and their formats) as a capability in the Initialize request and VS Code will create the UI based on this. 

In addition there would be a request to retrieve a range of objects from the adapter and we would need events to request updates of the UI from the backend.

/cc @ jacdavis @egamma @isidorn

**Comment by gregg-miskelly** [comment]
@weinand it doesn't affect engines that don't have modules - they simply never send any module load events and so the window is either empty (which is what we do in VS) or in an alternative implementation one could imagine the modules window is unavailable.

I definitely agree with you that if we were starting from scratch it might make sense to have no well known properties for modules, and let every debug adapter define its own set of properties. But if we want to host the protocol in VS that isn't going to work - we have a set of well known properties. So I have no problem if all the properties are optional (other than maybe 'Name', as that one seems pretty universal) as they are all optional to VS. But I think we should define these well known properties so that they can be mapped onto VS properties. If we want to also support other properties - sounds good to me.

**Comment by weinand** [comment]
I'm working on a detailed proposal...

**Comment by weinand** [comment]
@gregg-miskelly @jacdavis I've added a proposal what we should add to the debug protocol. Please review and provide feedback.
For the list of 'recommended' attributes of the `Module` type, I would like to add good names and descriptions so that potential adopters can understand their purpose (without having to understand VS).

**Comment by gregg-miskelly** [comment]
How flexible would you imagine 'ColumnDescriptor.width' being? Is this just a hint to help establish initial size of columns? Or would you expect it to be a hard limit? IMHO: it should be a hint. I would suggest we add that to the comment either way.

Suggested comments:

    /** Logical full path to the module. The exact definition is implementation defined, but usually this would be a full path to the on-disk file for the module */
    path?: string
    /** True if the module is optimized. */
    isOptimized?: bool
    /** True is the module is considered 'user code' by a debugger that supports JustMyCode. */
    isUseCode?: bool
    /** Version of Module. */
    version? : string
    /** User understandable description of if symbols were found for the module (ex: 'Symbols Loaded', 'Symbols not found', etc */
    symbolStatus?: string
    /** Logical full path to the symbol file. The exact definition is implementation defined. */
    symbolFilePath?: string
    /** Module created or modified? */
    dateTimeStamp?: string
    /** Address range covered by this module. */
    addressRange?: string

**Comment by jacdavis** [comment]
1) I believe the isUseCode is a typo. Should be isUserCode?

2) We'll likely also want to support actions. For instance, in the case of a module not having symbols loaded, we'd like to be able to throw the file picker. Upon selecting a file, this should send a request back to the adapter asking it to load symbols for a particular module at a specific path. The other case we would need for parity with VS is the ability to see the paths searched and the results of searching that path (file not found, symbol file doesn't match binary etc...)

3)  It does feel a little strange architecturally that adapter authors will be able to customize the columns in one case (vscode) and not in the other cases (VS and possibly Xamerin). We could certainly put a work item on our backlog to try and support custom columns inside VS, but it isn't clear where it would go in terms priority.

**Comment by weinand** [comment]
@gregg-miskelly Yes, ColumnDescriptor.width is a hint (and the details about the width specification have to be determined).

@gregg-miskelly @jacdavis I've added module support to the debug protocol with this commit:
https://github.com/Microsoft/vscode-debugadapter-node/commit/9a0d80c66fd36877be904b928d81a58ad00afb29
The change is available in version '1.10.0-pre.0' of the npm-modules 'vscode-debugprotocol' and 'vscode-debugadapter'.

@jacdavis
1. I'll fix `isUseCode` in the next revision
2. Yes, I see the need, so let's work on the details of protocol additions for the UI interactions (actions) after we've got a first implementation of the module view running in VS Code. Upfront design of this is difficult. Let's do it with some code in hand.
3. Yes, I agree but all contributing participants have to accept some compromises.
